# IDX Data Analyst Internship 

## Week 1
#### downloading the files, setting up python, setting up git, cleaning data.

In [5]:
#Imports
import pandas as pd
import crmls_listed
import crmls_sold


Total 31410 records exported to CRMLSListing202602.csv
Total 19106 records exported to CRMLSSold202602.csv


In [6]:
import glob

In [7]:
df_listings= pd.read_csv("CRMLSListing202602.csv")
df_sold = pd.read_csv("CRMLSSold202602.csv")

### Displaying the data frames 

In [8]:
glob.glob("*.csv")

['CRMLSListing202602.csv', 'CRMLSSold202602.csv']

In [9]:
len(glob.glob("CRMLSSold*.csv"))

1

In [10]:
sold_files = glob.glob("CRMLSSold*.csv")
listed_files = glob.glob("CRMLSListing*.csv")

print(len(sold_files))
print(len(listed_files))

1
1


In [11]:
import os
os.listdir()

['.DS_Store',
 'crmls_sold.py',
 'CRMLSListing202602.csv',
 '__pycache__',
 'README.md',
 '.gitignore',
 'CRMLSSold202602.csv',
 '.git',
 'crmls_listed.py',
 'analysis.ipynb',
 'raw']

In [12]:
sold_files = glob.glob("raw/CRMLSSold*.csv")
df_sold = pd.concat([pd.read_csv(f, low_memory= False) for f in sold_files], ignore_index=True)


In [13]:
df_listings = pd.concat(
    [pd.read_csv(f, low_memory=False) for f in listed_files],
    ignore_index=True)

In [14]:
# Looking at the original data set 
df_sold

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,...,PostalCode,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,latfilled,lonfilled,OriginatingSystemName,OriginatingSystemSubName,BuyerAgencyCompensationType,BuyerAgencyCompensation
0,Arcadia,Arcadia,NaN,False,NaN,NaN,False,1130000.0,538338723,espi.bagwell@kw.com,...,91016,0.0,14390.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Conejo,Conejo,NaN,True,NaN,NaN,NaN,1795000.0,535502637,david@mabryflp.com,...,93021,NaN,666032.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,OrangeCounty,OrangeCounty,NaN,True,NaN,NaN,NaN,99000.0,495549184,Jaredengberg@rocketmail.com,...,92283,0.0,871200.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,SouthwestRiversideCounty,SouthwestRiversideCounty,NaN,True,NaN,NaN,NaN,40000.0,446937009,Julie@whitehornhomes.com,...,92530,0.0,16553.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,HighDesert,HighDesert,NaN,True,NaN,NaN,NaN,59900.0,446912267,nadinesells4u@gmail.com,...,92327,0.0,275800.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
532590,NaN,NaN,NaN,NaN,NaN,NaN,NaN,649000.0,1008475397,chiggamasta424@gmail.com,...,90805,NaN,4682.0,NaN,NaN,NaN,NaN,NaN,Item1,2.5
532591,NaN,NaN,NaN,True,NaN,NaN,NaN,75689.0,1003434877,jhisquierdo@lee-associates.com,...,92371,NaN,946123.0,NaN,NaN,NaN,NaN,NaN,Item1,4.0
532592,NaN,NaN,NaN,True,NaN,NaN,NaN,275000.0,1002713752,MLS@DaltonRealtors.com,...,92082,0.0,619859.0,NaN,NaN,NaN,NaN,NaN,Item1,2.0
532593,NaN,NaN,NaN,False,NaN,NaN,False,10000.0,1002447928,Tazzie@CoronadoPremierProperties.com,...,92118,0.0,NaN,NaN,NaN,NaN,NaN,NaN,Item,100.0


In [15]:
#Looking at the original data set
df_listings

,OriginalListPrice,ListingKey,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,...,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,UnparsedAddress.1
0,1100000.0,1157171473,NaN,NaN,Cassandra,Diep,34.076588,-118.085541,3821 Earle Avenue,Residential,...,1.0,False,2.0,Rosemead,91770,NaN,0.0,7212.0,NaN,3821 Earle Avenue
1,2130000.0,1157170900,2026-04-09,2130000.0,Briant,Patterson,33.651866,-117.898412,2342 Westminster,Residential,...,6.0,False,2.0,Newport Mesa Unified,92627,Harcourts Prime Properties,0.0,8276.0,NaN,2342 Westminster
2,1199000.0,1157170763,NaN,NaN,Audrey,Huang,34.095744,-118.137994,312 N Marguerita Avenue,Residential,...,NaN,NaN,2.0,NaN,91801,NaN,NaN,7527.0,NaN,312 N Marguerita Avenue
3,530000.0,1157170224,2026-04-07,530000.0,Nick,Rogers,33.036771,-116.868485,708 12th street,Residential,...,NaN,False,2.0,NaN,92065,Coldwell Banker Realty,NaN,NaN,NaN,708 12th street
4,225000.0,1157169737,NaN,NaN,Tina,Quiroz,33.861935,-116.254086,0 Vacant Land,Land,...,NaN,False,NaN,NaN,92241,NaN,0.0,210395.0,NaN,0 Vacant Land
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
31405,4950.0,1149458718,NaN,NaN,Alphonso,Lascano,34.044402,-118.365074,1716 S Genesee Avenue,ResidentialLease,...,NaN,False,NaN,NaN,90019,NaN,NaN,4886.0,NaN,1716 S Genesee Avenue
31406,1699.0,1146439434,NaN,NaN,Gene,Smith,38.115612,-122.267146,124 Benson Ave,ResidentialLease,...,NaN,False,1.0,NaN,94590,NaN,NaN,15681.0,NaN,124 Benson Ave
31407,350000.0,1146104836,NaN,NaN,Jason,Lipovsky,33.139128,-117.053881,475 N Midway Drive 169,Residential,...,0.0,NaN,0.0,Escondido Union,92027,NaN,400.0,261360.0,NaN,475 N Midway Drive 169
31408,1325.0,1107246684,NaN,NaN,Kathy,Lund,32.572312,-117.034778,4589 Casa nova ct,ResidentialLease,...,5.0,False,2.0,Other,92154,NaN,0.0,2224.0,NaN,4589 Casa nova ct


#### Creating new data frames named df_sold_residential, and df_listings_residential, that contain only observations of residential homes. 

In [ ]:
# Looking at the property types in the data set before filtering
df_sold['PropertySubType'].value_counts()
# single family residence seems to be the largest sold property type

PropertySubType
SingleFamilyResidence    326253
Condominium               92138
Townhouse                 30770
Apartment                 11795
Duplex                    10090
ManufacturedOnLand         5427
Quadruplex                 3156
Triplex                    3104
StockCooperative           1802
MixedUse                   1615
Office                     1436
Retail                      952
Industrial                  612
Cabin                       445
Studio                      380
Business                    358
Warehouse                   266
RoomingHouse                246
SpecialPurpose              134
MultiFamily                 125
BoatSlip                    104
Agriculture                  78
UnimprovedLand               67
OwnYourOwn                   67
WaterPositionWithLand        64
Loft                         62
MobileHome                   55
ManufacturedHome             52
Timeshare                    17
CoOwnership                  16
Farm                    

In [ ]:
df_listings['PropertyType'].value_counts()
# residential homes are the largest listed property in the data. 

PropertyType
Residential            19997
ResidentialLease        5964
Land                    2226
ResidentialIncome       1318
ManufacturedInPark       938
CommercialSale           506
CommercialLease          354
BusinessOpportunity      107
Name: count, dtype: int64

In [50]:
# filtering csv files to property type == Residential.
df_sold_residential = df_sold[df_sold["PropertyType"]=="Residential"]
df_listings_residential = df_listings[df_listings["PropertyType"] == "Residential"]

In [45]:
# checking he data sets that have been filtered
df_sold_residential['PropertyType'].value_counts()

PropertyType
Residential    359471
Name: count, dtype: int64

In [46]:
df_listings_residential['PropertyType'].value_counts()

PropertyType
Residential    19997
Name: count, dtype: int64

#### Looking at the number of rows and columns in the data sets. 

In [36]:
df_listings.shape

(31410, 81)

In [37]:
df_sold.shape

(532595, 84)

In [43]:
df_listings_residential.shape

(19997, 81)

In [44]:
df_sold_residential.shape

(359471, 84)

Looking at the columns and the number of rows of the new and old data set. The original data sets are significantly larger compared to the new filtered data sets. 

df_listings has 31410 rows and 81 colums, df_sold has 532595 rows and 84 columns. 

I expect the new data frames to have the same amount of columns which it did, and the rows to reduce, 

df_listings_residential has 19997 rows and 81 columns, and df_sold_residential has 359471 rows and 84 columns. 

Deliverable – Week 1
Submit a .py script that concatenates all monthly files from January 2024 through the most recently
completed calendar month into two combined datasets (listings and sold), filters both to PropertyType ==
‘Residential’ only, and saves them as new CSVs. Include comments confirming row counts before and
after concatenation and before and after the Residential filter

#### Deliverable – Weeks 2–3
Submit a .py script documenting unique property types found, the filtering logic applied, and a null-count
summary table. Include a missing value report flagging any columns above 90% null. Produce a numeric
distribution summary (min, max, mean, median, percentiles) for ClosePrice, LivingArea, and
DaysOnMarket. Save the filtered dataset as a new CSV.

In [51]:
#Displaying column names to filter data and create new data frames.
df_sold_residential.columns

Index(['BuyerAgentAOR', 'ListAgentAOR', 'Flooring', 'ViewYN', 'WaterfrontYN',
       'BasementYN', 'PoolPrivateYN', 'OriginalListPrice', 'ListingKey',
       'ListAgentEmail', 'CloseDate', 'ClosePrice', 'ListAgentFirstName',
       'ListAgentLastName', 'Latitude', 'Longitude', 'UnparsedAddress',
       'PropertyType', 'LivingArea', 'ListPrice', 'DaysOnMarket',
       'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName',
       'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName',
       'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName',
       'FireplacesTotal', 'AssociationFeeFrequency', 'AboveGradeFinishedArea',
       'ListingKeyNumeric', 'MLSAreaMajor', 'TaxAnnualAmount',
       'CountyOrParish', 'MlsStatus', 'ElementarySchool', 'AttachedGarageYN',
       'ParkingTotal', 'BuilderName', 'PropertySubType', 'LotSizeAcres',
       'SubdivisionName', 'BuyerOfficeAOR', 'YearBuilt', 'StreetNumberNumeric',
       'ListingId', 'BathroomsTotalInteger', 'City', '

In [52]:
df_listings_residential.columns

Index(['OriginalListPrice', 'ListingKey', 'CloseDate', 'ClosePrice',
       'ListAgentFirstName', 'ListAgentLastName', 'Latitude', 'Longitude',
       'UnparsedAddress', 'PropertyType', 'LivingArea', 'ListPrice',
       'DaysOnMarket', 'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName',
       'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName',
       'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName',
       'FireplacesTotal', 'AssociationFeeFrequency', 'AboveGradeFinishedArea',
       'ListingKeyNumeric', 'MLSAreaMajor', 'TaxAnnualAmount',
       'CountyOrParish', 'PropertyType.1', 'MlsStatus', 'ElementarySchool',
       'ListAgentFirstName.1', 'AttachedGarageYN', 'ParkingTotal',
       'BuilderName', 'PropertySubType', 'LotSizeAcres', 'SubdivisionName',
       'BuyerOfficeAOR', 'YearBuilt', 'DaysOnMarket.1', 'StreetNumberNumeric',
       'LivingArea.1', 'ListingId', 'BathroomsTotalInteger', 'City', 'TaxYear',
       'BuildingAreaTotal', 'BedroomsTota

### Creatign new data frames with our variables of interest:

#### Key MLS Fields

List Price / Close Price — Transaction pricing data

Square Footage — Property size metrics

(LivingArea)
Bedrooms / Bathrooms — Property configuration

Property Type — Residential, Condo, etc.

Listing Agent / Brokerage — Agent and firm attribution

Days on Market (DaysOnMarket) — Time-to-sell metric

In [79]:
df_sold_new = df_sold_residential[["ClosePrice","ListPrice","BathroomsTotalInteger", "LotSizeSquareFeet","BedroomsTotal","PropertyType", "BuyerAgentAOR", "ListAgentAOR", "DaysOnMarket"]].copy()
df_sold_new.head()

,ClosePrice,ListPrice,BathroomsTotalInteger,LotSizeSquareFeet,BedroomsTotal,PropertyType,BuyerAgentAOR,ListAgentAOR,DaysOnMarket
0,1090000.0,1130000.0,4.0,14390.0,3.0,Residential,Arcadia,Arcadia,26
7,1995000.0,1995000.0,5.0,18295.0,4.0,Residential,CaliforniaDesert,CaliforniaDesert,0
8,2340000.0,2340000.0,3.0,11900.0,4.0,Residential,Laguna,Laguna,0
9,984000.0,984000.0,2.0,6500.0,3.0,Residential,PacificSouthwest,PacificSouthwest,0
10,1225000.0,1250000.0,4.0,5338.0,3.0,Residential,TheInlandGateway,TheInlandGateway,0


In [80]:
df_listing_new= df_listings_residential[["ListPrice","ClosePrice","BathroomsTotalInteger", "LotSizeSquareFeet","LivingArea", "BedroomsTotal","PropertyType", "DaysOnMarket", "ListAgentFirstName","ListAgentLastName", "ListAgentFullName","ListOfficeName","CoListOfficeName"]].copy()
df_listing_new.head()

,ListPrice,ClosePrice,BathroomsTotalInteger,LotSizeSquareFeet,LivingArea,BedroomsTotal,PropertyType,DaysOnMarket,ListAgentFirstName,ListAgentLastName,ListAgentFullName,ListOfficeName,CoListOfficeName
0,1100000.0,NaN,3.0,7212.0,2410.0,4.0,Residential,0,Cassandra,Diep,Cassandra Diep,eXp Realty of California Inc,eXp Realty of California Inc
1,2130000.0,2130000.0,4.0,8276.0,1903.0,6.0,Residential,0,Briant,Patterson,Briant Patterson,Pacific Sotheby's Int'l Realty,NaN
2,1199000.0,NaN,2.0,7527.0,1925.0,4.0,Residential,0,Audrey,Huang,Audrey Huang,COMPASS,NaN
3,530000.0,530000.0,1.0,NaN,1020.0,2.0,Residential,0,Nick,Rogers,Nick Rogers,Coldwell Banker Realty,NaN
5,1990000.0,NaN,5.0,7503.0,2557.0,5.0,Residential,0,Grace,Miranda,Grace Miranda,Keller Williams Realty World Media Center,Keller Williams Realty World Media Center


#### Summary statistics of the data frames:

In [81]:
df_listing_new.describe()

,ListPrice,ClosePrice,BathroomsTotalInteger,LotSizeSquareFeet,LivingArea,BedroomsTotal,DaysOnMarket
count,1.999700e+04,6.783000e+03,19997.000000,1.845100e+04,19986.000000,19988.000000,19997.000000
mean,1.345747e+06,1.233658e+06,2.660699,2.063192e+05,1978.072100,3.217080,23.836926
std,2.488062e+06,1.176476e+06,1.287351,9.445449e+06,1304.650215,1.186157,18.927245
min,2.900000e+04,2.850000e+04,0.000000,0.000000e+00,0.000000,0.000000,-10.000000
25%,5.749990e+05,6.445000e+05,2.000000,5.227000e+03,1248.000000,2.000000,8.000000
50%,8.300000e+05,9.100000e+05,2.000000,7.405000e+03,1688.000000,3.000000,20.000000
75%,1.385000e+06,1.460000e+06,3.000000,1.481020e+04,2337.000000,4.000000,39.000000
max,9.135000e+07,3.026000e+07,20.000000,1.079025e+09,47044.000000,17.000000,377.000000


In [82]:
df_sold_new.describe()

,ClosePrice,ListPrice,BathroomsTotalInteger,LotSizeSquareFeet,BedroomsTotal,DaysOnMarket
count,3.594690e+05,3.594710e+05,359404.000000,3.313540e+05,359462.000000,359471.000000
mean,1.188227e+06,1.139268e+06,2.530389,3.189625e+05,3.199114,36.297106
std,6.059275e+06,1.352650e+06,1.141028,1.562544e+07,1.065330,52.555171
min,0.000000e+00,5.250000e+02,0.000000,0.000000e+00,0.000000,-288.000000
25%,5.759030e+05,5.790000e+05,2.000000,5.227000e+03,3.000000,8.000000
50%,8.250000e+05,8.190000e+05,2.000000,7.232000e+03,3.000000,18.000000
75%,1.300000e+06,1.295000e+06,3.000000,1.176100e+04,4.000000,47.000000
max,9.895000e+08,1.375000e+08,175.000000,2.208318e+09,45.000000,12430.000000


the summary statistics of both the data sets are similar when we look at the the mean, median, and the range of the data. the std deviation varies more in the df_sold_new data set, as compared to the df_listings-new data set, which is to be expected as generally properties sold are closed at a lower price than the list. 

Looking at the min and max, in the listing data there is a min value of -10 days on market, which is concerning and needs further inspection. Similar observations can be seen for the sold data set with a value of -288. The max value of Days on market of the sold data being 12430 is also concerning because that is very far from the average days on market.  

### Cleaning the data 
##### 1) Looking at the missing values, which to remove and which to keep.
##### 2) changing any quatitative variable with letters inside. 

In [83]:
df_sold_new.isna().sum()


ClosePrice                   2
ListPrice                    0
BathroomsTotalInteger       67
LotSizeSquareFeet        28117
BedroomsTotal                9
PropertyType                 0
BuyerAgentAOR            46179
ListAgentAOR             46179
DaysOnMarket                 0
dtype: int64

#### Calculating the percent of null values in the data 


In [92]:
(df_sold_new.isnull().mean() *100).round(4)

ClosePrice                0.0006
ListPrice                 0.0000
BathroomsTotalInteger     0.0186
LotSizeSquareFeet         7.8218
BedroomsTotal             0.0025
PropertyType              0.0000
BuyerAgentAOR            12.8464
ListAgentAOR             12.8464
DaysOnMarket              0.0000
dtype: float64

In [93]:
(df_listing_new.isnull().mean()*100).round(4)

ListPrice                 0.0000
ClosePrice               66.0799
BathroomsTotalInteger     0.0000
LotSizeSquareFeet         7.7312
LivingArea                0.0550
BedroomsTotal             0.0450
PropertyType              0.0000
DaysOnMarket              0.0000
ListAgentFirstName        0.6251
ListAgentLastName         0.0000
ListAgentFullName         0.0250
ListOfficeName            0.0000
CoListOfficeName         76.9715
dtype: float64

In [27]:
df_listing_new.isna().sum()

ListPrice                   90
ClosePrice               20753
LotSizeSquareFeet         2671
LivingArea                4273
BedroomsTotal             4156
BathroomsTotalInteger     2887
PropertyType                 0
DaysOnMarket                 0
ListAgentFirstName         154
ListAgentLastName            1
ListAgentFullName           17
ListOfficeName               0
CoListOfficeName         25034
dtype: int64

git add .
git commit -m "your message"
git pull origin main --rebase
git push origin main